### DISCUSSION TITLE
为什么 TCP 协议有粘包问题 · Why's THE Design? · /whys-the-design-tcp-message-frame

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#598] [Announcements]
https://draveness.me/whys-the-design-tcp-message-frame 

 粘包并不是 TCP 协议造成的，它的出现是因为应用层协议设计者对 TCP 协议的错误理解，忽略了 TCP 协议的定义并且缺乏设计应用层协议的经验。本文将从 TCP 协议以及应用层协议出发，分析我们经常提到的 TCP 协议中的粘包是如何发生的。很多人可能会认为粘包是一个比较低级的甚至不值得讨论的问题，但是在作者看来这个问题还是很有趣的，不是所有人都系统性地学过基于 TCP 的应用层协议设计，也不是所有人对 TCP 协议也没有那么深入的理解，相信很多人学习编程的过程都是自底向上的，所以作者认为这是一个值得回答的问题，我们应该传递正确的知识，而不是负面的和居高临下的情绪。

**Comment by yuyongID** [comment]
处理TCP流分包是做私有协议序列化和反序列化的比较令人头疼的问题。

**Comment by draveness** [comment]
> 处理TCP流分包是做私有协议序列化和反序列化的比较令人头疼的问题。

不过这也是基于 TCP 的应用层协议必须做的😂

**Comment by iostalks** [comment]
UPD 是有消息边界的，所以不会出现粘包。应用层的数据包大小应该尽可肯能接近 UPD 的最大负荷，可以充分利用带宽。

如果应用层数据包超过 UPD 的最大负荷，不知道是不是就丢包了~~

**Comment by draveness** [comment]
> UPD 是有消息边界的，所以不会出现粘包。应用层的数据包大小应该尽可肯能接近 UPD 的最大负荷，可以充分利用带宽。
> 
> 如果应用层数据包超过 UPD 的最大负荷，不知道是不是就丢包了~~

UDP 协议的上限是 65535，因为长度位存储的数据有限，但是一般都会被路径 MTU 限制，可以看 https://draveness.me/whys-the-design-tcp-segment-ip-packet

**Comment by KielChan** [comment]
我是用golang解析jmx的时候遇到这个问题，RMI协议，没有长度也没有终止符，吐了；听说netty里面的做法是不停的轮询加载，unmarshal，直到超时；

**Comment by draveness** [comment]
> 我是用golang解析jmx的时候遇到这个问题，RMI协议，没有长度也没有终止符，吐了；听说netty里面的做法是不停的轮询加载，unmarshal，直到超时；

这也是个办法.....

**Comment by yuyongID** [comment]
> @KielChan
> 我是用golang解析jmx的时候遇到这个问题，RMI协议，没有长度也没有终止符，吐了；听说netty里面的做法是不停的轮询加载，unmarshal，直到超时；

我用go解析私有协议的时候也是这样的，不停轮询，看能不能反序列包头，再拆包头反序列化包体长度值跟收到的包体的是否一致。然后取出包体，再尝试反序列化。一顿操作猛如虎，只为取出消息体。

**Comment by draveness** [comment]
> > @KielChan
> > 我是用golang解析jmx的时候遇到这个问题，RMI协议，没有长度也没有终止符，吐了；听说netty里面的做法是不停的轮询加载，unmarshal，直到超时；
> 
> 我用go解析私有协议的时候也是这样的，不停轮询，看能不能反序列包头，再拆包头反序列化包体长度值跟收到的包体的是否一致。然后取出包体，再尝试反序列化。一顿操作猛如虎，只为取出消息体。

为啥不设计终止符或者长度，这其实不符合基于 TCP 的协议设计规范..

**Comment by yuyongID** [comment]
> @draveness
> > > @KielChan
> > > 我是用golang解析jmx的时候遇到这个问题，RMI协议，没有长度也没有终止符，吐了；听说netty里面的做法是不停的轮询加载，unmarshal，直到超时；
> > 
> > 我用go解析私有协议的时候也是这样的，不停轮询，看能不能反序列包头，再拆包头反序列化包体长度值跟收到的包体的是否一致。然后取出包体，再尝试反序列化。一顿操作猛如虎，只为取出消息体。
> 
> 为啥不设计终止符或者长度，这其实不符合基于 TCP 的协议设计规范..

有长度的，只是长度也需要反序列化才知道是否正确。

**Comment by draveness** [comment]
> > @draveness
> > > > @KielChan
> > > > 我是用golang解析jmx的时候遇到这个问题，RMI协议，没有长度也没有终止符，吐了；听说netty里面的做法是不停的轮询加载，unmarshal，直到超时；
> > > 
> > > 
> > > 我用go解析私有协议的时候也是这样的，不停轮询，看能不能反序列包头，再拆包头反序列化包体长度值跟收到的包体的是否一致。然后取出包体，再尝试反序列化。一顿操作猛如虎，只为取出消息体。
> > 
> > 
> > 为啥不设计终止符或者长度，这其实不符合基于 TCP 的协议设计规范..
> 
> 有长度的，只是长度也需要反序列化才知道是否正确。

一般情况下都是消息的前几个字节会表示长度，不需要反序列化的

**Comment by dawncold** [comment]
“HTTPz 头”

---

*2020-05-15 UPDATES：已修复*

**Comment by Crearns** [comment]
你好，想请教一下UDP的边界是怎么确定的？是根据长度确定边界吗？

**Comment by draveness** [comment]
> 你好，想请教一下UDP的边界是怎么确定的？是根据长度确定边界吗？

每个 UDP 数据包都表示一个完整的消息

**Comment by TigerZhag** [comment]
“面向字节流”往上面数第四行 “也不是所有人对 TCP 协议也没有那么深入的理解”，病句🙅🙅


---

*2020-07-12 UPDATES: 已修复*

**Comment by zerouwar** [comment]
很nice的文章，对TCP有新的理解

**Comment by icbd** [comment]
课后题:

1. UDP 是面向 Datagram 的, 每个 UDP 都是独立的. 为了不被 IP 因为 MTU 限制而分片, 需要注意控制 UDP payload 不大于 
`MTU - IP_HEADER_20 - UDP_HEADER_8`. 
  UDP 不会出现粘包问题.

2. 基于长度分帧: MySQL; 基于终结符分帧: Redis.

**Comment by monkeyWie** [comment]
请教一下如果UDP的包大小超过MTU限制，导致在ip层分片，那么接收方是不是一样要做沾包处理呢？

**Comment by mrdear** [comment]
> @monkeyWie
> 请教一下如果UDP的包大小超过MTU限制，导致在ip层分片，那么接收方是不是一样要做沾包处理呢？

接收方IP层会做数据合并,接收到时,已经是一个完整的消息了,所以不存在粘包现象.

**Comment by hydrogenwuyan** [comment]
您好，有个问题想请教一下。如果并发的向一个socket写数据，会不会出现数据混在一起导致服务端解析出来的消息达不到预期呢？如果不是，tcp是如何来进行控制的呢？

**Comment by draveness** [comment]
> 您好，有个问题想请教一下。如果并发的向一个socket写数据，会不会出现数据混在一起导致服务端解析出来的消息达不到预期呢？如果不是，tcp是如何来进行控制的呢？

UDP 多线程写不会出问题，TCP 不应该这么设计，总的来说要避免多线程同时写

**Reply by EricQi7** [reply]
为什么udp多线程写不会出问题呢？不也会出现数据混在一起的情况吗

**Comment by tabris17** [comment]
「粘包」建议打引号

**Comment by hehedahehe** [comment]
挺不错的

**Comment by leoleedev** [comment]
很好的文章，解答了自己对于 HTTP协议如何确定消息边界的疑惑
